File: Pico-8SEG-LED/python/shu.py
import spidev
import time
import os



KILOBIT   = 0xFE
HUNDREDS  = 0xFD
TENS      = 0xFB
UNITS     = 0xF7
Dot       = 0x80



SEG8Code = [
    0x3F, # 0
    0x06, # 1
    0x5B, # 2
    0x4F, # 3
    0x66, # 4
    0x6D, # 5
    0x7D, # 6
    0x07, # 7
    0x7F, # 8
    0x6F, # 9
    0x77, # A
    0x7C, # b
    0x39, # C
    0x5E, # d
    0x79, # E
    0x71  # F
    ] 




class GPIO:
    INPUT = "in"
    OUTPUT = "out"
    HIGH = "1"
    LOW = "0"

    def __init__(self, pin):
        self.pin = pin
        self.export()

    def export(self):
        try:
            with open("/sys/class/gpio/export", "w") as f:
                f.write(str(self.pin))
        except OSError as e:
            if e.errno == 16:  # Resource busy
                print(f"GPIO {self.pin} is already exported.")

    def set_mode(self, mode):
        with open(f"/sys/class/gpio/gpio{self.pin}/direction", "w") as f:
            f.write(mode)

    def write(self, value):
        with open(f"/sys/class/gpio/gpio{self.pin}/value", "w") as f:
            f.write(value)

    def read(self):
        with open(f"/sys/class/gpio/gpio{self.pin}/value", "r") as f:
            return f.read().strip()

    def cleanup(self):
        try:
            with open("/sys/class/gpio/unexport", "w") as f:
                f.write(str(self.pin))
        except OSError:
            print(f"Failed to unexport GPIO {self.pin}")

RCLK_PIN = GPIO(449)#B1


class LED_8SEG():
    def __init__(self):
        RCLK_PIN.set_mode(GPIO.OUTPUT)
        RCLK_PIN.write(GPIO.HIGH)
        self.spi = spidev.SpiDev()
        self.spi.open(0, 0)  # 打开 SPI 端口 0, 设备 0
        self.spi.max_speed_hz = 1000_000
        self.spi.mode = 0b00
        self.SEG8=SEG8Code
       
   
        
    '''
    function: Send Command
    parameter: 
        Num: bit select
        Seg：segment select       
    Info:The data transfer
    '''
    def write_cmd(self, Num, Seg):    
        RCLK_PIN.write(GPIO.HIGH)
        self.spi.xfer2(bytearray([Num]))
        self.spi.xfer2(bytearray([Seg]))
        RCLK_PIN.write(GPIO.LOW)
        time.sleep(0.002)
        RCLK_PIN.write(GPIO.HIGH)

    def cleanup(self):
        self.spi.close()
        RCLK_PIN.cleanup()
        
if __name__=='__main__':
    LED = LED_8SEG()
    #color BRG
    for o in range(99999):
            time.sleep(0.0005)
            LED.write_cmd(UNITS,LED.SEG8[o%10])
            time.sleep(0.0005)
            LED.write_cmd(TENS,LED.SEG8[(o%100)//10])
            time.sleep(0.0005)
            LED.write_cmd(HUNDREDS,LED.SEG8[(o%1000)//100]|Dot)
            time.sleep(0.0005)
            LED.write_cmd(KILOBIT,LED.SEG8[(o%10000)//1000])

    
    LED.cleanup()

        

File: Pico-8SEG-LED/c/shu.c
#include "shu.h"
#include <wiringx.h>
#include "unistd.h"

#define KILOBIT   0XFE //KiloBit
#define HUNDREDS  0XFD //hundreds
#define TENS      0XFB //tens
#define UNITS     0XF7 //units
#define Dot       0x80 //decimal separator

#define LED_8SEG_RCLK_PIN  50
int number = 0;
void main()
{

DEV_Module_Init();



while(1)
{

LED_8SEG_stopwatch();
//Increment();// 个位0-9

}





}


uint8_t SEG8Code[] ={
        0x3F, // 0
        0x06, // 1
        0x5B, // 2
        0x4F, // 3
        0x66, // 4
        0x6D, // 5
        0x7D, // 6
        0x07, // 7
        0x7F, // 8
        0x6F, // 9
        0x77, // A
        0x7C, // b
        0x39, // C
        0x5E, // d
        0x79, // E
        0x71, // F
    };

/******************************************************************************
function:	Send Command
parameter: 
    Num: bit select
    Seg：segment select
Info:The data transfer
******************************************************************************/

void LED_8SEG_SendCommand(uint8_t Num, uint8_t Seg) 
{
    digitalWrite(LED_8SEG_RCLK_PIN, HIGH);
    DEV_SPI_WriteByte(Num);
    DEV_SPI_WriteByte(Seg);
    digitalWrite(LED_8SEG_RCLK_PIN, LOW);
    sleep(0.002);
    digitalWrite(LED_8SEG_RCLK_PIN, HIGH);
}

/******************************************************************************
function:	Stop watch
parameter: 
Info:99 seconds watch
******************************************************************************/

void LED_8SEG_stopwatch()
{
    
    for (int o = 0; o < 99999; o++)
    {
        sleep(0.8);
        LED_8SEG_SendCommand(UNITS, SEG8Code[o%10]);
        sleep(0.8);
        LED_8SEG_SendCommand(TENS, SEG8Code[(o%100)/10]);
        sleep(0.8);
        LED_8SEG_SendCommand(HUNDREDS, SEG8Code[(o%1000)/100] | Dot);
        sleep(0.8);
        LED_8SEG_SendCommand(KILOBIT, SEG8Code[(o%10000)/1000]);
    }
    
     //时间可以自己进行调试
    
   


}

void Increment()
{
        LED_8SEG_SendCommand(UNITS, SEG8Code[number]);
        sleep(1);

        number +=1;

        if(number>9)

        number=0;


}